cpp/dolfinx/fem/assemble_scalar_impl.h

```diff
@@ -273,31 +211,47 @@ T assemble_scalar(
         md::mdspan<const T, md::extents<std::size_t, md::dynamic_extent, 2,
                                         md::dynamic_extent>>(
             coeffs.data(), facets.size() / shape1, 2, cstride),
-        perms, cdofs_b);
+        facet_perms, cdofs_b);
   }
 
-  for (int i = 0; i < M.num_integrals(IntegralType::vertex, 0); ++i)
+  for (fem::IntegralType itg_type :
+       {fem::IntegralType::exterior_facet, fem::IntegralType::vertex,
+        fem::IntegralType::ridge})
   {
-    auto fn = M.kernel(IntegralType::vertex, i, 0);
-    assert(fn);
-
-    auto& [coeffs, cstride] = coefficients.at({IntegralType::vertex, i});
-
-    std::span<const std::int32_t> vertices
-        = M.domain(IntegralType::vertex, i, 0);
-    assert(vertices.size() * cstride == coeffs.size());
-
-    constexpr std::size_t num_adjacent_cells = 1;
-    // Two values per adj. cell (cell index and local vertex index).
-    constexpr std::size_t shape1 = 2 * num_adjacent_cells;
-
-    value += impl::assemble_vertices(
-        x_dofmap, x,
-        md::mdspan<const std::int32_t,
-                   md::extents<std::size_t, md::dynamic_extent, 2>>(
-            vertices.data(), vertices.size() / shape1, shape1),
-        fn, constants,
-        md::mdspan(coeffs.data(), vertices.size() / shape1, cstride), cdofs_b);
+    md::mdspan<const std::uint8_t, md::dextents<std::size_t, 2>> perms;
+    switch (itg_type)
+    {
+    case fem::IntegralType::exterior_facet:
+    {
+      perms = facet_perms;
+      break;
+    }
+    default:
+      break;
+    }
+
+    for (int i = 0; i < M.num_integrals(itg_type, 0); ++i)
+    {
+      auto fn = M.kernel(itg_type, i, 0);
+      assert(fn);
+      auto& [coeffs, cstride] = coefficients.at({itg_type, i});
+
+      std::span entities = M.domain(itg_type, i, 0);
+
+      constexpr std::size_t num_adjacent_cells = 1;
+      // Two values per each adj. cell (cell index and local entity index).
+      constexpr std::size_t shape1 = 2 * num_adjacent_cells;
```

Comment:
This is trivial - why is it needed?

Reply:
I didn't add this. It exists in the previous code. Can remove it.